Fix behavior of --no-signature to turn signing off

koterpillar opened the pull request:
The --no-signature switch should enable signing rather than disable it;
consequently, the default for 'stack upload' should be to sign the package.

Fixes #3739.

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

koterpillar commented:
I've started writing an integration test, but realized I can't have it upload anything to (real) Hackage. Is there something I'm missing that would let me test this properly?

dbaynard commented:
I wonder whether [https://hackage.haskell.org/upload#candidates](the package candidates) feature is suitable for this.

koterpillar commented:
This took longer than I expected, but...

I've made the Hackage URL configurable and added an integration test that tests uploading with and without signing.

Would such a test be useful?

koterpillar commented on src/Stack/Upload.hs:
```diff
@@ -126,13 +126,14 @@ applyCreds creds req0 = do
 --
 -- Since 0.1.2.1
-uploadBytes :: HackageCreds
+uploadBytes :: String -- ^ Hackage base URL
```
Yes, it is, for the library but not the executable. Does it need to be documented somewhere?

dbaynard replied:
Perhaps in the changelog. I'll see what the other reviewer says.

I was pointing it out as it means this will be merged for the next major release (which is due very soon) — I'm just trying to maintain good habits.

koterpillar commented on ChangeLog.md:
```diff
@@ -40,6 +40,9 @@ Behavior changes:
   [#4133](https://github.com/commercialhaskell/stack/issues/4133).
+* `stack upload` signs the package by default, as documented. `--no-signature`
+  turns the signing off.
+  [#3739](https://github.com/commercialhaskell/stack/issues/3739)
```
This is fixing the regression introduced in f7275f8795c6be26ab0819f1e8bc010d2e575f6e, when the `--no-signature` was accidentally flipped around.

dbaynard replied:
Oh, yes. It may need to be made very prominent — especially if people have been relying on this behaviour.

koterpillar replied:
The other option is to document the existing default and apply `--sign` option to `upload` as well as `sdist`.

dbaynard replied:
I think what you've already done is great. I might move it higher up the changelog before release.

dbaynard commented:
Thanks, @koterpillar, the changes are good. Just waiting for another reviewer (for reasons described above). It should be done soon.

dbaynard commented:
Thanks, @koterpillar — good job.

borsboom commented:
@koterpillar The `upload` integration test seems to be hanging on Windows (it never finishes).  Do you have access to a Windows machine by any chance to help debug?

dbaynard commented on test/integration/tests/upload/files/FakeHackage.hs:
```diff
@@ -0,0 +1,8 @@
+import Network.Simple.TCP
+
+-- | Fake server that always responds with HTTP OK
+main =
```
Try `main = withSocketsDo $`

dbaynard commented:
https://github.com/haskell/network/blob/8bdf58ef59862b3b09e378b8133b74301ebc4eda/Network/Socket/Internal.hs#L207-L228

koterpillar commented:
I'm trying to get a Windows environment to test on. Will post once I have any progress.

koterpillar commented:
Here's the list of things I had to change:

- `withSocketsDo`
- It's not really good to send a HTTP response before a request was received
- Somehow terminating the process doesn't work well, so I added a timeout
- Added `gpg.bat` to PATH because Windows doesn't see a `gpg` without an extension

Now it works at least on my VM.
